### PR TITLE
All space infos

### PR DESCRIPTION
**Pull request by SimonRichardson**
## Description of change

This lifts the `MapWithSpaceNames` requirement of state out of the function. Instead, it requires a lookup of `network.SpaceInfos` to determine if and how to map ProviderIDs to SpaceNames.

The great thing with this is that we can now prevent excessive calls to state when requesting the space names for the bindings.

Future work should attempt to lift this from out of state to make in conceptually pure. This has no requirements on state, it transforms one piece of data from one to another, and isn't relatable to what's in the database.

## QA steps

Ensure that the tests pass. This is mainly mechanical change, but existing tests should work and pass as previously had done.

### CODE REVIEW

**Comment by SimonRichardson** [comment]
https://trello.com/c/Qg6jjSrT/2607-handle-mapwithspacenames-fetching-allspaceinfos-correctly

**Comment by SimonRichardson** [comment]
!!build!!

Intermittent network issues.

File: apiserver/facades/client/client/status.go [hunk 1]
```diff
@@ -743,6 +743,15 @@ func fetchAllApplicationsAndUnits(
 		}
 	}
 
+	// Get all the space infos before iterating over the application infos.
```
Comment (manadart) [inline apiserver/facades/client/client/status.go:746]: This is more of a patch comment as opposed to something I would leave in the code-base. The code alone is explanation enough.

This is a nice gain, but a further (micro) optimisation would be to acquire it even higher in the call stack and pass it to `fetchNetworkInterfaces` as well as this method.

File: apiserver/facades/client/bundle/bundle.go [hunk 2]
```diff
@@ -651,15 +664,15 @@ func (b *BundleAPI) printSpaceNamesInEndpointBindings(apps []description.Applica
 	return false
 }
 
-func (b *BundleAPI) endpointBindings(bindings map[string]string, printValue bool) (map[string]string, error) {
+func (b *BundleAPI) endpointBindings(bindings map[string]string, allSpacesInfoLookup network.SpaceInfos, printValue bool) (map[string]string, error) {
```
Comment (manadart) [inline apiserver/facades/client/bundle/bundle.go:667]: `%s/allSpacesInfoLookup/spaceLookup/g`?

File: apiserver/facades/client/client/status.go (hunk 1 above)
Comment (SimonRichardson) [reply apiserver/facades/client/client/status.go:746]: I did try this, but it does cause a lot of changes. I'll make a comment about that as a further optimization we can lift this up further.

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**